### PR TITLE
Updates to deal with 2.0.0 tag issues

### LINKED ISSUE #53: ODS 2.x review
I just reviewed the latest ODS version (July 17) as well as the corresponding tables and have some questions/comments (I already sent some of them to @gleckler1, but maybe they got lost):

1. Note 6: Remove dataset_version_number as it is now included in source_id? Or use source_label instead of source_id?
2. grid_label: Is it "gr[i]" as described in the table or "g[i]" as described in Note 8?
3. mip_era: In the ODS this is fixed to "obs4MIPs", but in the json tables it is "CMIP5/6". Which one is correct?
4. source_id: According to the ODS, source_id must include the version number. Does that mean a user has to register a separate source for each dataset version?
5. source_version_number is missing in the ODS global attributes table
6. ncdump of example file should be updated
7. Filename: Remove source_version_number as it is already included in source_id? Or use source_label instead of source_id?

### LINKED ISSUE #54: output_path_template and output_file_template are reversed in the demo
These are inverted, compared to the [google doc](https://docs.google.com/document/d/1FXXBhUh71Hjus557ZTD3EKPi_2zxeLvi1aICXOjVYPc/edit#heading=h.1t3h5sf)
```
"output_path_template": "<variable_id><table_id><source_id><variant_label><grid_label><version>",
"output_file_template": "<activity_id><institution_id><source_id><table_id><variable_id><grid_label>",
```
This should be
```
"output_path_template": "<activity_id><institution_id><source_id><table_id><variable_id><grid_label><version>",
"output_file_template": "<variable_id>_<table_id>_<source_id>_<variant_label>_<grid_label>[_<time_range>].nc",
```

### LINKED ISSUE #57: demo: obs4MIPs_CV.json corrections
Add following to "required_global_attributes":
institution  #registered
region #registered
source #registered in `source_id`
source_label #registered in `source_id`
variant_label

Remove following from "required_global_attributes":  source_version_number, mip_era

Remove "mip_era" from repo

Include in `obs4MIPs_CV.json`:
"activity_id": "obs4MIPs"

Remove coordinate and grid entries from `obs4MIPs_CV.json` (they are obtained by CMOR from a different json file).

### LINKED ISSUE #58: demo: rssSsmiPrw-input.json corrections
This file should be replaced with the attached.
[obs4MIPs_input_json.docx](https://github.com/PCMDI/obs4MIPs-cmor-tables/files/1163614/obs4MIPs_input_json.docx)

### LINKED ISSUE #59: obs4MIPs_CVs.json  change in "source_id" list
I suggest structuring the source_id listing in this table as follows:
```
"source_id":{
        "REMSS-PRW-6-6-0":{
            "description":"Precipitable Water",
            "institution_id":"RSS",
            "label":"REMSS PRW v6.6.0"
            "release_year":"2017",
            "source_id":"REMSS-PRW-6-6-0",
            "source_label":"REMSS-PRW",
            "source_type":"satellite_blended",
            "region":"global"
        },
}
```

This will then be stored in `obs4MIPs_CV.json` (rewritten by `writeJson.py`) as:
```
            "source":"REMSS PRW v6.6.0 (2017): Precipitable Water",
```
If we collected this information before a dataset gotten written, then given a source_id, CMOR could automatically generate the following global attributes: institution_id, institution, source_label, source, source_type, and region.  They wouldn't have to be set by the user in the obs4MIPs_input.json file.

We need to be sure that source_type and region are invariably implied by source_id (and can't have multiple values).  (If multiple values were allowed, then the file names would not necessarily be unique, I think ... well in the case of "region" I suppose the "grid_label" could differ.)

### PR DESCRIPTION
**Pull request by durack1**
@taylor13 @dnadeau4 @gleckler1 
Fix #53 
Fix #54 
Fix #57 
Fix #58 
Fix #59 

This is a work in progress PR, due to the `CMOR` error described in PCMDI/cmor#209

***NOTE: DEMO IS NOT CURRENTLY WORKING DUE TO THE ERROR LINKED ABOVE***

Remaining issues to resolve:
- [x] `mip_era`
- [x] `further_info_url` (This requires CMOR code changes, should land in 3.2.6)
- [x] `obs4MIPs_` table names
- [x] `tracking_id`
- [x] `data_specs_version`
- [x] `source_id`

### CODE REVIEW

**Comment by durack1** [comment]
There is still code to implement, once the `cmor` error is resolved - see [src/writeJson.py#L650-L652](https://github.com/PCMDI/obs4MIPs-cmor-tables/blob/5d13fbedabdfa278e992d49fc59672ebcf61c096/src/writeJson.py#L650-L652)

**Comment by durack1** [comment]
@dnadeau4 @taylor13 do I have these settings correct?

See [demo/rssSsmiPrw-input.json#L43-L45](https://github.com/PCMDI/obs4MIPs-cmor-tables/blob/5d13fbedabdfa278e992d49fc59672ebcf61c096/demo/rssSsmiPrw-input.json#L43-L45)

**Comment by gleckler1** [comment]
@durack1 @taylor13 @dnadeau4  PD this is looking very nice and clear.   I wonder if
 "contact": "PCMDI (pcmdi-cmip@llnl.gov)",

is a good idea... it might be misleading.   

I suggest we use the same contact that is in the Tech Note of the dataset:   support@remss.com

**Comment by durack1** [comment]
@taylor13 @gleckler1 @dnadeau4 this is getting closer:
```
// global attributes:
		:Conventions = "CF-1.7 ODS-2.0" ;
		:activity_id = "obs4MIPs" ;
		:contact = "RSS (support@remss.com)" ;
		:creation_date = "2017-07-24T22:16:14Z" ;
		:data_specs_version = "01.00.13" ;
		:data_structure = "grid" ;
		:external_variables = "areacella" ;
		:frequency = "mon" ;
		:further_info_url = "https://furtherinfo.es-doc.org/<activity_id><institution_id><source_label><variable_id>" ;
		:grid = "2.5x2.5 degree latitude x longitude" ;
		:grid_label = "gn" ;
		:history = "2017-07-24T22:16:14Z CMOR rewrote data to be consistent with CF standards and CMIP6 requirements." ;
		:institution = "Remote Sensing Systems, Santa Rosa, CA 95401, USA" ;
		:institution_id = "RSS" ;
		:mip_era = "CMIP6" ;
		:nominal_resolution = "250 km" ;
		:product = "observations" ;
		:realm = "atmos" ;
		:region = "global" ;
		:source = "REMSS PRW v6.6.0 (2017): Precipitable water (PRW)" ;
		:source_id = "REMSS-PRW-6-6-0" ;
		:source_label = "REMSS-PRW" ;
		:source_type = "satellite_blended" ;
		:table_id = "Amon" ;
		:table_info = "Creation Date:(21 July 2017) MD5:2d3f9890f664a68044fdf2309a91d50d" ;
		:title = "PRW V6.6.0 prepared for obs4MIPs (ODS-v2.0.0)" ;
		:tracking_id = "46a836e2-e958-4033-b899-4e54418cb1d4" ;
		:variable_id = "prw" ;
		:variant_info = "Best Estimate" ;
		:variant_label = "BE" ;
		:license = "Data in this file produced by <Your Centre Name> is licensed under a Creative Commons Attribution-ShareAlike 4.0 International License (https://creativecommons.org/licenses/). Use of the data must be acknowledged following guidelines found at <a URL maintained by you>. Further information about this data, including some limitations, can be found via <some URL maintained by you>" ;
		:cmor_version = "3.2.5" ;
}
```

**Comment by durack1** [comment]
@taylor13 @gleckler1 @dnadeau4 the standing issues with the above (demo output file) are:

* `:mip_era = "CMIP6" ;` should not appear
* `:further_info_url = "http://furtherinfo.es-doc.org/<activity_id><institution_id><source_label><variable_id>" ;` should be `:further_info_url = "http://furtherinfo.es-doc.org/obs4MIPs.RSS.REMSS-PRW.prw" ;`
* `:tracking_id = "37acf47f-7363-4384-a7f4-5febf5e063c0" ;` should be `:tracking_id = "hdl:21.14102/37acf47f-7363-4384-a7f4-5febf5e063c0" ;`

**Comment by taylor13** [comment]
I spoke with Denis.  He is modifying the code to allow further_info_url to be set from the user_input.json file, even if activity is *not* "CMIP6"

further_info_url is an "internal" variable, so in the .json file set _further_info_url (with initial underscore).

The "tracking_prefix" seems to work for any project in Denis' test suite.  Are you sure its spelled correctly in the user_input.json file?

**Comment by durack1** [comment]
@taylor13 @dnadeau4 great, I indeed had a problem with the `input.json` which had `tracking_id` rather than `tracking_prefix`. After correcting this I now get:
`:tracking_id = "hdl:21.14102/b12eadb7-14b1-4c24-bd1a-72ac07880ef6" ;
` in the file

I'll await the `_further_info_url` fix tomorrow, it'd be great to get this demo finalized

**Comment by durack1** [comment]
@taylor13 @dnadeau4 ok great, this is where I've managed to get up to:
```
// global attributes:
		:Conventions = "CF-1.7 ODS-2.0" ;
		:activity_id = "obs4MIPs" ;
		:contact = "RSS (support@remss.com)" ;
		:creation_date = "2017-07-26T01:59:52Z" ;
		:data_specs_version = "2.0.0" ;
		:data_structure = "grid" ;
		:external_variables = "areacella" ;
		:frequency = "mon" ;
		:further_info_url = "http://furtherinfo.es-doc.org/<activity_id><institution_id><source_label><variable_id>" ;
		:grid = "2.5x2.5 degree latitude x longitude" ;
		:grid_label = "gn" ;
		:history = "2017-07-26T01:59:52Z CMOR rewrote data to be consistent with CF standards and  requirements." ;
		:institution = "Remote Sensing Systems, Santa Rosa, CA 95401, USA" ;
		:institution_id = "RSS" ;
		:nominal_resolution = "250 km" ;
		:product = "observations" ;
		:realm = "atmos" ;
		:region = "global" ;
		:source = "REMSS PRW v6.6.0 (2017): Precipitable water (PRW)" ;
		:source_id = "REMSS-PRW-6-6-0" ;
		:source_label = "REMSS-PRW" ;
		:source_type = "satellite_blended" ;
		:table_id = "Amon" ;
		:table_info = "Creation Date:(25 July 2017) MD5:2d3f9890f664a68044fdf2309a91d50d" ;
		:title = "PRW V6.6.0 prepared for obs4MIPs (ODS-v2.0.0)" ;
		:tracking_id = "hdl:21.14102/a7edf9c2-e119-4ba2-8520-647dd6709ab9" ;
		:variable_id = "prw" ;
		:variant_info = "Best Estimate" ;
		:variant_label = "BE" ;
		:license = "Data in this file produced by <Your Centre Name> is licensed under a Creative Commons Attribution-ShareAlike 4.0 International License (https://creativecommons.org/licenses/). Use of the data must be acknowledged following guidelines found at <a URL maintained by you>. Further information about this data, including some limitations, can be found via <some URL maintained by you>" ;
		:cmor_version = "3.2.5" ;
}
```
I note that now I am getting the following `CMOR` warning (which is fine):
```
(cmor325) bash-3.2$ python runCmorDemo.py 

C Traceback:
In function: cmor_set_dataset_att
! called from: cmor_load_table_internal
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Warning: table: , This keyword: dataRequest_specs_version value (table: , This keyword: dataRequest_specs_version) is not a valid table header entry.!
 Use the user input JSON file to add custom attributes.
!
!!!!!!!!!!!!!!!!!!!!!!!!!

! ------
! CMOR is now closed.
! ------
! During execution we encountered:
!   1 Warning(s)
!   0 Error(s)
! ------
! Please review them.
! ------
!
```
I'm also no longer able to name my `obs4MIPs` tables as I like, with the existing entry:
```
{
    "Header":{
...
        "table_id":"Table obs4MIPs_Amon"
    },
```
I get the following error:
```
!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: The attribute "table_id" could not be validated. 
! The current input value is "obs4MIPs_Amon" which is not valid 
! Valid values must match the regular expression:
! 	["^Aday$" "^Amon$" "^Lmon$" "^Omon$" "^SImon$" "^fx$"  ...] 
! 
! Check your Control Vocabulary file ".//obs4MIPs_CV.json".
! 
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```

**Comment by durack1** [comment]
`source_id` also needs amending, as `obs4MIPs_CV.json` contains the component `source` entry:
```
 {
     "source_id":{
        "REMSS-PRW-6-6-0":{
            "description":"Precipitable Water",
            "institution_id":"RSS",
            "label":"REMSS PRW v6.6.0",
            "region":"global",
            "release_year":"2017",
            "source":"REMSS PRW v6.6.0 (2017): Precipitable water (PRW)",
            "source_id":"REMSS-PRW-6-6-0",
            "source_label":"REMSS-PRW",
            "source_type":"satellite_blended"
         }
     }
}
```
Should become:
```
{
     "source_id":{
        "REMSS-PRW-6-6-0":{
            "description":"Precipitable Water",
            "institution_id":"RSS",
            "label":"REMSS PRW v6.6.0",
            "region":"global",
            "release_year":"2017",
            "source_id":"REMSS-PRW-6-6-0",
            "source_label":"REMSS-PRW",
            "source_type":"satellite_blended"
         }
     }
 }
```

File: demo/rssSsmiPrw-input.json [hunk 1]
```diff
@@ -1,35 +1,47 @@
...
-    "output_path_template":         "<variable_id><table_id><source_id><variant_label><grid_label><version>",
-    "output_file_template":         "<activity_id><institution_id><source_id><table_id><variable_id><grid_label>",
+    "#comment3": "The following should not be edited by the user",
 
-    "license":                      "Data in this file produced by <Your Centre Name> is licensed under a Creative Commons Attribution-ShareAlike 4.0 International License (https://creativecommons.org/licenses/). Use of the data must be acknowledged following guidelines found at <a URL maintained by you>. Further information about this data, including some limitations, can be found via <some URL maintained by you>"
-}
\ No newline at end of file
+    "activity_id":                  "obs4MIPs",
+    "product":                      "observations",
+    "further_info_url":            "http://furtherinfo.es-doc.org/<activity_id><institution_id><source_label><variable_id>",
```
Comment (durack1) [inline demo/rssSsmiPrw-input.json]: Update these to:
```
http://furtherinfo.es-doc.org/<activity_id><institution_id><source_label><source_id><variable_id>
https://furtherinfo.es-doc.org/obs4MIPs.RSS.REMSS-PRW.REMSS-PRW-6-0-0.prw
```

Comment (dnadeau4) [inline demo/rssSsmiPrw-input.json]: Can you defined this in your JSON input file.
"_further_info_url_tmpl"

example:
    "_further_info_url_tmpl":        "http://furtherinfo.es-doc.org/<activity_id><institution_id><source_label><source_id><variable_id>",

File: Tables/obs4MIPs_Aday.json [hunk 2]
```diff
@@ -3,13 +3,13 @@
...
+        "dataRequest_specs_version":"01.00.13",
+        "data_specs_version":"2.0.0",
         "generic_levels":"",
-        "mip_era":"CMIP6",
         "missing_value":"1e20",
         "product":"observations",
         "realm":"atmos",
-        "table_date":"17 July 2017",
+        "table_date":"26 July 2017",
         "table_id":"Aday"
```
Comment (durack1) [inline Tables/obs4MIPs_Aday.json]: This table should have a `obs4MIPs ` prefix

**Comment by durack1** [comment]
@dnadeau4 not quite - this is using `cmor: 2017.07.26.3.2.6-np19py27h5cd32f7_0 pcmdi/label/nightly`
```
(cmor326Nightly) bash-3.2$ python runCmorDemo.py 
/Users/durack1/anaconda2/envs/cmor326Nightly/lib/python2.7/site-packages/cdms2/axis.py:769: UserWarning: genutil module not present, was not able to determine if axis is level based on units
  "genutil module not present, was not able to determine if axis is level based on units")

C Traceback:
! In function: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: Your Control Vocabulary file specifies one or more
! required attributes.  The following
! attribute was not properly set.
! 
! Please set attribute: "further_info_url" in your input file.
!
!!!!!!!!!!!!!!!!!!!!!!!!!


C Traceback:
! In function: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: Please fix required attributes mentioned in
! the warnings/error above and rerun. (aborting!)
! 
!
!!!!!!!!!!!!!!!!!!!!!!!!!


C Traceback:
! In function: 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: Cannot continue until you fix the errors listed above: -1
!
!!!!!!!!!!!!!!!!!!!!!!!!!

Traceback (most recent call last):
  File "runCmorDemo.py", line 55, in <module>
    cmor.write(varid,values,time_vals=time[:],time_bnds=time.getBounds()) ; # Write variable with time axis
  File "/Users/durack1/anaconda2/envs/cmor326Nightly/lib/python2.7/site-packages/cmor/pywrapper.py", line 795, in write
    time_vals, time_bnds, store_with)
_cmor.CMORError: Problem with 'cmor.write'. Please check the logfile (if defined).
```

**Comment by taylor13** [comment]
It's quite clear that a fatal error has been encountered, but it would be nice to be told what is expected as a global attribute and what was actually provided.  Can that information be easily added to the error message?  (Not highest priority, however).

**Comment by durack1** [comment]
@dnadeau4 @taylor13 @gleckler1 this now seems finalized:
```
// global attributes:
		:Conventions = "CF-1.7 ODS-2.0" ;
		:activity_id = "obs4MIPs" ;
		:contact = "RSS (support@remss.com)" ;
		:creation_date = "2017-07-27T01:35:44Z" ;
		:data_specs_version = "2.0.0" ;
		:data_structure = "grid" ;
		:external_variables = "areacella" ;
		:frequency = "mon" ;
		:further_info_url = "http://furtherinfo.es-doc.org/obs4MIPs.RSS.REMSS-PRW.REMSS-PRW-6-6-0.prw" ;
		:grid = "2.5x2.5 degree latitude x longitude" ;
		:grid_label = "gn" ;
		:history = "2017-07-27T01:35:44Z CMOR rewrote data to be consistent with CF standards and CMIP6 requirements." ;
		:institution = "Remote Sensing Systems, Santa Rosa, CA 95401, USA" ;
		:institution_id = "RSS" ;
		:mip_era = "CMIP6" ;
		:nominal_resolution = "250 km" ;
		:product = "observations" ;
		:realm = "atmos" ;
		:region = "global" ;
		:source = "REMSS PRW v6.6.0 (2017): Precipitable water (PRW)" ;
		:source_id = "REMSS-PRW-6-6-0" ;
		:source_label = "REMSS-PRW" ;
		:source_type = "satellite_blended" ;
		:table_id = "obs4MIPs_Amon" ;
		:table_info = "Creation Date:(26 July 2017) MD5:f428bf8092725094454eccffc5875fa7" ;
		:title = "PRW V6.6.0 prepared for obs4MIPs (ODS-v2.0.0)" ;
		:tracking_id = "hdl:21.14102/e642f8e8-5bdb-4805-a13d-adaeb1b49460" ;
		:variable_id = "prw" ;
		:variant_info = "Best Estimate" ;
		:variant_label = "BE" ;
		:license = "Data in this file produced by <Your Centre Name> is licensed under a Creative Commons Attribution-ShareAlike 4.0 International License (https://creativecommons.org/licenses/). Use of the data must be acknowledged following guidelines found at <a URL maintained by you>. Further information about this data, including some limitations, can be found via <some URL maintained by you>" ;
		:cmor_version = "3.2.6" ;
}
```
I will merge this tomorrow if none of you have any concerns..

@dnadeau4 thanks for these fixes in `CMOR3.2.6`!!

**Comment by durack1** [comment]
@dnadeau4 I can confirm this works on both `OS X` and `Linux`, I tested using the latest nightlies

**Comment by taylor13** [comment]
The ncdump should be included in  https://goo.gl/jVZsQl

**Comment by durack1** [comment]
@gleckler1 ping:

The ncdump should be included in https://goo.gl/jVZsQl - the latest ncdump is found at https://github.com/PCMDI/obs4MIPs-cmor-tables/pull/60#issuecomment-318232328

**Comment by durack1** [comment]
@gleckler1 here is the complete ncdump of a freshly written demo file:
[prw_obs4MIPs_Amon_REMSS-PRW-6-6-0_BE_gn_198701-198811.txt](https://github.com/PCMDI/obs4MIPs-cmor-tables/files/1181035/prw_obs4MIPs_Amon_REMSS-PRW-6-6-0_BE_gn_198701-198811.txt)

**Comment by gleckler1** [comment]
@durack1 ok I'll update doc with this

**Pull request merged**